### DISCUSSION TITLE
Render canvas-on-canvas annotations

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#17] [How Might We]
From the developer point of view this is the same as #1, #2, #4 etc. You won't necessarily know that canvas A has canvas B painted onto it (almost certainly amongst other content directly painted).

In any canvas-on-canvas scenario (for spatial and temporal), the component needs to translate all annotations on canvas B into the coordinate space of Canvas A, and then treat them the same as any other annotations on Canvas A.

Scenario:

Two canvases representing pages of books. Each canvas has text annotations.
I create a new, synthetic canvas for a 2-up view, paint the two canvases onto it, and render the new synthetic canvas using the component. It behaves exactly as if all the text annotations had been made directly on the synthetic canvas.

AV scenario (where canvas-on-canvas came from originally)
Two chunks of video, each with text transcripts (captions), each represented by a canvas
Sequence them together so that one follows the other directly on the new canvas.
(may not get to `duration` yet).

**Comment by tomcrane** [comment]
Proposal


```html
<canvas-panel id="cp"></canvas-panel>
<script>

   const cp = document.getElementById("cp");
   const vault = cp.vault; // see #49
   await vault.loadManifest("https://iiif.wellcomecollection.org/presentation/b18035723");

   // make a 2-page spread
   let lhs = vault.FromRef("https://iiif.wellcomecollection.org/presentation/b18035723/canvases/b18035723_0007.JP2");
   let rhs = vault.FromRef("https://iiif.wellcomecollection.org/presentation/b18035723/canvases/b18035723_0008.JP2");

   // these are close enough not to require anything fancy done...
   console.log(lhs.width, lhs.height); // 2008, 2736
   console.log(rhs.width, rhs.height); // 2008, 2740

   const bigCanvas = new Canvas("https://example.org/synthetic-id");
   
   let lhsTarget = new Target();
   lhsTarget.spatial.x = 0;
   lhsTarget.spatial.y = 0;
   lhsTarget.spatial.w = lhs.width;
   lhsTarget.spatial.h = lhs.height;
   
   let rhsTarget = new Target();
   rhsTarget.spatial.x = lhs.width + 10; // a gutter of 10 canvas-units between them
   rhsTarget.spatial.y = 0;
   rhsTarget.spatial.w = rhs.width;
   rhsTarget.spatial.h = rhs.height;

   bigCanvas.paint(lhs, lhsTarget);
   bigCanvas.paint(rhs, rhsTarget);
   bigCanvas.setDimensionsFromContent(); // a utility method; will set bigCanvas.width and bigCanvas.height

   vault.load(bigCanvas); // should vault take the id from this by default?
   cp.setCanvas(bigCanvas.id);

</script>  
```

**Reply by tomcrane** [reply]
See also the purely declarative Atlas example here:

https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/48

This is probably an easier way to achieve the visual equivalent - but it's not the same as a canvas on canvas...

If you wanted to scale, align, etc - you might want to do it the manual way shown earlier.

**Comment by stephenwf** [comment]
A limitation currently would be canvas + mixed media on the same canvas. These cases would have to be handled in a different way, perhaps by wrapping the "extra" media in another canvas.

**Comment by tomcrane** [comment]
The example in https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/17#discussioncomment-1169238 is being used to achieve 2-up - but that's just one application of canvas on canvas annotations.

For the Canvas Panel API, what actually happens?
In that example we have a couple of canvases already tracked by Vault:

```
   let lhs = vault.FromRef("https://iiif.wellcomecollection.org/presentation/b18035723/canvases/b18035723_0007.JP2");
   let rhs = vault.FromRef("https://iiif.wellcomecollection.org/presentation/b18035723/canvases/b18035723_0008.JP2");
```

We then create a new Canvas by instantiating... what? This isn't yet a vault-tracked Canvas... What library does the Canvas class live in?

```
  const bigCanvas = new Canvas("https://example.org/synthetic-id");
```

We then see a couple of functions on our Canvas class being called:

```
   bigCanvas.paint(lhs, lhsTarget);
   bigCanvas.paint(rhs, rhsTarget);
   bigCanvas.setDimensionsFromContent(); // a utility method; will set bigCanvas.width and bigCanvas.height
```

Where are these functions defined? Is that how they work?
(`.paint(..)` would be used to associate media with the canvas, too - thinking about Manifest editor scenario, the "drop" of an image onto the canvas would be a `paint(..)` and then it might be moved around.)

Now we give the new Canvas to Vault to track:

```
   vault.load(bigCanvas); // should vault take the id from this by default?
   cp.setCanvas(bigCanvas.id);
```

Is this sketchy API the way we want to do this?
